cmd/result-aggregator/main.go

```diff
@@ -0,0 +1,119 @@
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	errorRate = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "ci_operator_error_rate",
+			Help: "number of errors, sorted by label/type",
+		},
+		[]string{"job_name", "type", "state", "reason", "cluster"},
+	)
+
+	port string
+)
+
+func init() {
+	flag.StringVar(&port, "port", "8080", "Port the server listens to")
+
+	prometheus.MustRegister(errorRate)
+}
+
+// Request holds the data used to report a result to an aggregation server
+type Request struct {
+	// JobName is the name of the job for which a result is being reported
+	JobName string `json:"job_name"`
+	// Type is the type of job ("presubmit", "postsubmit", "periodic" or "batch")
+	Type string `json:"type"`
+	// State is "succeeded" or "failed"
+	State string `json:"state"`
+	// Reason is a colon-delimited list of reasons for failure
+	Reason string `json:"reason"`
+
+	Cluster string `json:"cluster"`
+}
+
+func validateRequest(request *Request) error {
+	if request.Reason == "" {
+		return fmt.Errorf("reason field in request is empty")
+	}
+	if request.JobName == "" {
+		return fmt.Errorf("job_name field in request is empty")
+	}
+	if request.State == "" {
+		return fmt.Errorf("state field in request is empty")
+	}
+	if request.Type == "" {
+		return fmt.Errorf("type field in request is empty")
+	}
+	if request.Cluster == "" {
+		return fmt.Errorf("cluster field in request is empty")
+	}
+	return nil
+}
+
+func handleError(w http.ResponseWriter, err error) {
+	w.WriteHeader(http.StatusBadRequest)
+	fmt.Fprint(w, err)
+}
+
+func withErrorRate(request *Request) {
+	labels := prometheus.Labels{
+		"job_name": request.JobName,
+		"type":     request.Type,
+		"state":    request.State,
+		"reason":   request.Reason,
+		"cluster":  request.Cluster,
+	}
+	errorRate.With(labels).Inc()
+}
+
+func handleCIOperatorResult() http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		bytes, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			handleError(w, fmt.Errorf("unable to ready request body: %v", err))
+			return
+		}
+
+		request := &Request{}
+		if err := json.Unmarshal(bytes, request); err != nil {
+			handleError(w, fmt.Errorf("unable to decode request body: %v", err))
+			return
+		}
+
+		if err := validateRequest(request); err != nil {
+			handleError(w, err)
+			return
+		}
+
+		withErrorRate(request)
+
+		w.WriteHeader(http.StatusOK)
+
+		logrus.Infof("Request with %#v request processed", request)
+	})
+}
+
+func main() {
```

Comment:
mind using the `interrupts` package from Prow to start the server, provide health + readiness endpoints and serve prom metrics?

Reply:
How do I access the metrics with `interrupts`?